feat: add automated plant watering workshop for educators (#41)
viam-devrel/viamcodelabs: site/viamguides/src/workshop-plant-watering/workshop-plant-watering.md

   - **Determine extensibility**: During the Hands-on Experiment portion of the delivery plan, there are many options to guide the workshop on a different learning path. Determine if this section will be self-guided exploration, a group exercise, or self-paced learning for participants to explore on their own after the workshop.
   - **Prepare code and documentation**: Review the provided code and add comments as necessary.
   - **Review discussion topics**: Review the suggested group discussion topics and adjust for your participants.
   - **Review quiz questions**: Review the suggested quiz questions and adjust for your participants. Determine if knowledge assessment will be completed throughout the workshop or afterwards.
1. **Dry Run and Testing**
   - **Test the workshop project**: Run through the build following the prepared documentation to ensure it works as expected. Simulate common issues and prepare troubleshooting tips for these scenarios. If physical space is limited, consider sharing supplies.
   - **Bring backups**: In addition to the backup supplies mentioned earlier during planning.

<!-- ------------------------ -->

## Delivery Plan - overview

Duration: 2

> Feel free to make a copy of [this sample slide deck](https://docs.google.com/presentation/d/1BUkv3h-49KYPi-HqHn36P38V67ThM9-UVU-muOQZ0GU/edit#slide=id.g314410d34af_0_25), customize it, and make it your own.

Reviewer: ```suggestion
Feel free to make a copy of [this sample slide deck](https://docs.google.com/presentation/d/1_mI__LEwgJApM_3Z2yKwaWOjCPyVPndVNHyrybrDt2g/edit?usp=sharing), customize it, and make it your own.
```

1. Connect your Pi to Wi-Fi so that you can run `viam-server` wirelessly. Check **Configure wireless LAN** and enter your wireless network credentials. SSID (short for Service Set Identifier) is your Wi-Fi network name, and password is the network password. Change the section `Wireless LAN country` to where your router is currently being operated.
   ![raspberry pi hostname username and password](assets/generalSettings.png)
1. Select the **SERVICES** tab, check **Enable SSH**, and select **Use password authentication**.
   ![raspberry pi enable SSH](assets/sshSettings.png)
   > aside negative
   > Be sure that you remember the `hostname` and `username` you set, as you will need this when you SSH into your Pi.
1. **Save** your updates, and confirm `YES` to apply OS customization settings. Confirm `YES` to erase data on the microSD card. You may also be prompted by your operating system to enter an administrator password. After granting permissions to the Imager, it will begin writing and then verifying the Linux installation to the microSD card.
1. Remove the microSD card from your computer when the installation is complete.

<form>

>   <name>WWhat is the purpose of setting a custom `hostname` during the Raspberry Pi OS configuration?</name>

Reviewer: ```suggestion
  <name>What is the purpose of setting a custom `hostname` during the Raspberry Pi OS configuration?</name>
```

   ![add board](assets/addBoard.png)
1. Notice adding this module adds the board hardware component called `board-1`. The collapsible card on the right corresponds to the part listed in the left sidebar.
   ![added board](assets/boardAdded.png)
1. Click **Save** in the top right to save and apply your configuration changes.
   > aside negative
   > If any problems occur, check under the **LOGS** tab to see what might be going wrong.

<form>
  <name>How does adding modules like `board-1` contribute to the overall architecture of your machine?</name>
  <input type="radio" value="They abstract hardware-specific details, allowing software to interact with components through a consistent API.">
  <input type="radio" value="They provide direct power and connectivity to the components.">
  <input type="radio" value="They eliminate the need for the Raspberry Pi in the system.">
  <input type="radio" value="They store the configuration locally on the air sensor.">
</form>

<!-- ------------------------ -->

## Delivery Plan - Hands-on Experiment

Duration: 40

### Test the moisture sensor

> 1. In [the Viam app](https://app.viam.com/fleet/locations) under the **CONTROL** tab, you can see the hardware component called `board-1``. That's your Raspberry Pi that you configured on the **CONFIGURE** tab.

Reviewer: ```suggestion
1. In [the Viam app](https://app.viam.com/fleet/locations) under the **CONTROL** tab, you can see the hardware component called `board-1`. That's your Raspberry Pi that you configured on the **CONFIGURE** tab.
```

   ![control tab](assets/controlTab.png)
1. The module we added has a `GetGPIO` function to get the digital output of a pin. Enter `40` in the "Pin" field, select the "Read" mode, and click "Get" for the "State" section under "READ PIN". "Low" means there is enough moisture, "High" means the moisture level has fallen below the threshold:
   ![read GPIO](assets/readGPIO.png)

   > aside negative
   > The moisture threshold is adjusted by turning the screw on the top of the sensor module. As you get into the Control interface of the Viam app, you can refine the threshold based on the needs of your plant.

   > aside negative
   > If any problems occur, check under the **LOGS** tab to see what might be going wrong.

### Test the pump

> 1. In [the Viam app](https://app.viam.com/fleet/locations) under the **CONTROL** tab, you can see the hardware component called `board-1``. That's your Raspberry Pi that you configured on the **CONFIGURE** tab.

Reviewer: ```suggestion
1. In [the Viam app](https://app.viam.com/fleet/locations) under the **CONTROL** tab, you can see the hardware component called `board-1`. That's your Raspberry Pi that you configured on the **CONFIGURE** tab.
```

1.  Find the corresponding card to `process-1`. Enter the executable path of Python3 running on your Raspberry Pi that you output from a previous step. Add an argument of the `process.py` file to run on your Raspberry Pi. Enter the working directory where you want the process to execute.
    ![configure process](assets/configureProcess.png)
1.  Still within the `process-1` card, select the advanced settings icon near the top right corner to review the configuration JSON. Create a new `env` property, and add your environment variables within the new property, formatted like the following with your own credentials.
    ```json
      "env": {
        "BOARD_NAME": "board-1",
        "ROBOT_API_KEY": "your-api-key",
        "ROBOT_API_KEY_ID": "your-api-key-id",
        "ROBOT_ADDRESS": "your-robot-address"
      },
    ```
    ![configure JSON](assets/configureJSON.png)
    > aside negative

>     > The `BOARD_NAME` is the default names for the Raspberry Pi when added to our Viam machine. Other machine credentials can be found under the **CONNECT** tab, selecting an SDK, and toggling **Include API key** to reveal your credentials within the code sample.

Reviewer: ```suggestion
    > The `BOARD_NAME` is the default name for the Raspberry Pi when added to our Viam machine. Other machine credentials can be found under the **CONNECT** tab, selecting an SDK, and toggling **Include API key** to reveal your credentials within the code sample.
```

- How can the design of the device be improved for better accessibility, performance, or durability?

### Quiz questions ❓

Review the suggested quiz questions below to evaluate participants' understanding of key concepts. These questions can be used to assess knowledge either during (see multiple choice answers in the delivery plan) or after the workshop.

1. **Hardware Assembly**
   - What is the purpose of setting a custom `hostname` during the Raspberry Pi OS configuration?
   - What hardware signals are used to indicate with the soil moisture is too low?
1. **Software Setup**
   - Why is it important to connect your machine to the Viam app during setup?
   - What does it mean that `viam-server` enables remote access to the Raspberry Pi?
   - How does the Viam app simplify the process of testing and controlling hardware components like the board?
1. **Hands-on Experiment**
   - What is the primary role of `viam-server` in the system architecture of your plant watering setup?
   - What is the purpose of the `env` field in your process configuration?

### Next-level projects

> Here are some ideas for expanding and enhancing the current project to create autonomous behaviors.

Reviewer: Consider adding a link to this project (Nicolai's Viam video): https://youtu.be/l4MQRiHPY6Y?feature=shared
